non_prov_records.go

```diff
@@ -0,0 +1,193 @@
+package dht
+
+import (
+	"context"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	u "github.com/ipfs/go-ipfs-util"
+	"github.com/jbenet/goprocess"
+	"github.com/jbenet/goprocess/context"
+	"github.com/libp2p/go-libp2p-kad-dht/providers"
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+)
+
+// vars for cleaning up expired records
+var nonProvRecordCleanupInterval = time.Hour
+
+// maxNonProvRecordAge specifies the maximum time that any node will hold onto a record
+// from the time its received. This does not apply to any other forms of validity that
+// the record may contain.
+// For example, a record may contain an ipns entry with an EOL saying its valid
+// until the year 2020 (a great time in the future). For that record to stick around
+// it must be rebroadcasted more frequently than once every 'maxNonProvRecordAge'
+var maxNonProvRecordAge = time.Hour * 36
+
+// vars for republishing records
+var nonProvRecordRePublishInterval = 1 * time.Hour
+var nonProvRecordRePublishAge = 1 * time.Hour
+var enableRepublishJitter = true
+
+type NonProvRecordsManager struct {
+	dht *IpfsDHT
+	ctx context.Context
+
+	proc   goprocess.Process
+	dstore ds.Batching
+
+	cleanupInterval time.Duration // scan interval for expiring records
+
+	rePublishInterval time.Duration // scan interval for republishing records
+}
+
+func NewNonProvRecordsManager(ctx context.Context, dht *IpfsDHT, dstore ds.Batching) *NonProvRecordsManager {
+	m := new(NonProvRecordsManager)
+	m.dht = dht
+	m.ctx = ctx
+	m.dstore = dstore
+	m.proc = goprocessctx.WithContext(ctx)
+
+	// expire records beyond maxage
+	m.cleanupInterval = nonProvRecordCleanupInterval
+	m.proc.Go(m.expire)
+
+	// republish records older than prescribed age
+	m.rePublishInterval = nonProvRecordRePublishInterval
+	m.proc.Go(m.rePublish)
+
```

Comment:
@bigs I think we can have separate dedicated go-routines for expiry & republish as we don't need to serialize access to a resource/synchronize  between them. Let me know what you think.